### PR TITLE
Admin "Reload Station" Map Generator - Reloads specified block to roundstart. also buildmode area mapgens have names instead of paths on a list.

### PR DESCRIPTION
**Pull request by silicons**
nonantag evaders maxcapped so many times in the past week i feel this is necessary.
I'm trying to make this work properly, as I had to improvise with the cropped maploading.
:cl:
rscadd: Badmins: Buildmode map generators have names in the list to select them, instead of paths.
rscadd: Also, a new map generator has been added, repair/reload station. Use it VERY sparingly, it deletes the block of the map and reloads it to roundstart. THIS CAN CAUSE ISSUES WITH MACHINES AND ATMOSPHERICS, SO DO NOT USE IT UNLESS YOU ABSOLUTELY HAVE TO!
experimental: The reload station one tagged DO NOT USE shouldn't be used as it doesn't delete anything before loading, so if you use it you'll have two copies of things. That can result in a LOT of issues, so don't use it unless you're a codermin and know what you're doing/abusing!
/:cl:

### CODE REVIEW

**Comment by kevinz000** [comment]
ok this is just 100% broken, how the fuck do I get the goddamn maploader to selectively load?

**Comment by kevinz000** [comment]
maploader cropping working (inefficient as fuck, it loops through everything and skips it instead of loading it if it's nto in the crop range, instead of only going through the ones that need loading, I don't know how to do the latter right.)

atoms loaded won't init for some reason.

File: code/modules/mapping/reader.dm [hunk 1]
```diff
@@ -29,21 +29,23 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
...
+	. = load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf, lower_crop_x, upper_crop_x, lower_crop_y, upper_crop_y)
 	#ifdef TESTING
 	if(turfsSkipped)
 		testing("Skipped loading [turfsSkipped] default turfs")
 	#endif
 	Master.StopLoadingMap()
 
-/dmm_suite/proc/load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf)
+/dmm_suite/proc/load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf, x_lower = 0, x_upper = INFINITY, y_lower = 0, y_upper = INFINITY)
+	to_chat(world, "<span class='notice'>Maploader: Loading [dmm_file] with offsets [x_offset]/[y_offset]/[z_offset] with cropping set to [cropMap], measureOnly set to [measureOnly], \
```
Comment (RemieRichards) [inline code/modules/mapping/reader.dm]: ...

File: code/modules/mapping/reader.dm [hunk 2]
```diff
@@ -29,21 +29,23 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
  * 2) Read the map line by line, parsing the result (using parse_grid)
  *
  */
-/dmm_suite/load_map(dmm_file as file, x_offset as num, y_offset as num, z_offset as num, cropMap as num, measureOnly as num, no_changeturf as num)
+/dmm_suite/load_map(dmm_file as file, x_offset as num, y_offset as num, z_offset as num, cropMap as num, measureOnly as num, no_changeturf as num, lower_crop_x as num, upper_crop_x as num, lower_crop_y as num, upper_crop_y as num)
```
Comment (RemieRichards) [inline code/modules/mapping/reader.dm]: x,y,x,y is standard, not x,x,y,y

**Comment by kevinz000** [comment]
This now considers lighting correctly/works without lagging as hard as before (still inefficient but better.)

**Comment by WJohn** [comment]
This is really cool. It's like regenerating the thunderdome but for everything!

**Comment by JamieH** [comment]
So can I mark 2 corners and then it'll regenerate that area? :D

**Comment by kevinz000** [comment]
@JamieH it'll delete everything BUT the floor and players if it doesn't break on you.
there's some runtimes with maploading but it isn't connected to my code so :P 
tl;dr don't use this unless you absolutely have to but this will more or less restore a chunk to roundstart state without deleting hte players. if you have an alternative, use it, but if someone bombed the station to shit, this is your tool.

File: code/controllers/subsystem/air.dm [hunk 3]
```diff
@@ -71,6 +71,8 @@ SUBSYSTEM_DEF(air)
 
 
 /datum/controller/subsystem/air/fire(resumed = 0)
+	if(map_loading)
+		return					//Runtimes will occur if things are still being initialized.
```
Comment (Cyberboss) [inline code/controllers/subsystem/air.dm]: There's already handling for this you meme

File: code/modules/mapping/reader.dm [hunk 4]
```diff
@@ -60,6 +62,9 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
 	var/key_len = 0
 
 	var/stored_index = 1
+
+	to_chat(usr, "<span class='notice'>Maploader initialized. Loop processing.</span>")
```
Comment (Cyberboss) [inline code/modules/mapping/reader.dm]: >[READY]

File: code/modules/mapping/reader.dm [hunk 5]
```diff
@@ -79,9 +84,15 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
 		// (1,1,1) = {"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"}
 		else if(dmmRegex.group[3]) // Coords
 			if(!key_len)
+				to_chat(usr, "<span class='warning'>ABORT: Coords before model definition in DMM")
```
Comment (Cyberboss) [inline code/modules/mapping/reader.dm]: .

**Review by Cyberboss — DISMISSED** [review]

**Comment by Tacolizard** [comment]
so what fluff will be used to explain this feature to players so we don't break their emmulsions?

File: code/modules/mapping/reader.dm (hunk 4 above)
Comment (kevinz000) [inline code/modules/mapping/reader.dm]: It only goes to user which is admin who invoked it

**Comment by kevinz000** [comment]
@Tacolizard like I give a fuck :^)

File: code/modules/mapping/reader.dm (hunk 5 above)
Comment (lordpidey) [inline code/modules/mapping/reader.dm]: You're missing a `</span>

**Comment by lordpidey** [comment]
"Nanotrasen Centcom Alert:

Your station has been involved in a horrific bluespace teleportation accident involving another, identical spacestation.  The good news is noone in your station was harmed.  There are no survivors from the other station."

**Comment by kevinz000** [comment]
fixed

**Comment by kevinz000** [comment]
@Cyberboss reee i fixed it the usr to_chats are for feedback for the admin who invoked it

File: code/modules/mapping/reader.dm (hunk 4 above)
Comment (RemieRichards) [inline code/modules/mapping/reader.dm]: ```usr``` is not acceptable.

**Review by RemieRichards — DISMISSED** [review]
```usr``` is not acceptable.

File: code/modules/mapping/reader.dm [hunk 6]
```diff
@@ -29,21 +29,23 @@ GLOBAL_DATUM_INIT(_preloader, /dmm_suite/preloader, new)
...
-	. = load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf)
+	. = load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf, lower_crop_x, upper_crop_x, lower_crop_y, upper_crop_y)
 	#ifdef TESTING
 	if(turfsSkipped)
 		testing("Skipped loading [turfsSkipped] default turfs")
 	#endif
 	Master.StopLoadingMap()
 
-/dmm_suite/proc/load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf)
+/dmm_suite/proc/load_map_impl(dmm_file, x_offset, y_offset, z_offset, cropMap, measureOnly, no_changeturf, x_lower = -INFINITY, x_upper = INFINITY, y_lower = -INFINITY, y_upper = INFINITY)
```
Comment (Jalleo) [inline code/modules/mapping/reader.dm]: add a user input into the proc usr output of any kind is bad

**Comment by optimumtact** [comment]
>maploader cropping working (inefficient as fuck, it loops through everything and skips it instead of loading it if it's nto in the crop range, instead of only going through the ones that need loading, I don't know how to do the latter right.)

This hurts me :(

**Comment by kevinz000** [comment]
O

**Comment by PKPenguin321** [comment]
Fix this NOW

**Comment by kevinz000** [comment]
I AM I AM

File: code/modules/mapping/reader.dm (hunk 6 above)
Comment (kevinz000) [inline code/modules/mapping/reader.dm]: ?

**Comment by kevinz000** [comment]
fixed

**Comment by optimumtact** [comment]
https://github.com/tgstation/tgstation/pull/28986/files?w=1

You need to undo all the pointless whitespace changes you're doing, and configure your editor to stop rewriting line endings

**Comment by kevinz000** [comment]
ok

**Comment by kevinz000** [comment]
fixed
